<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN"
"http://www.w3.org/TR/html4/strict.dtd">
<html>
	<head>
		<script type="text/javascript" src="../../js/jquery-1.4.2.min.js"></script>
		<script type="text/javascript" src="../../js/local.js"></script>
		<!-- Google Analytics tracking code-->
		<script type="text/javascript">
			var _gaq = _gaq || [];
			_gaq.push(['_setAccount', 'UA-16521993-1']);
			_gaq.push(['_trackPageview']);

			(function() {
				var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
				ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
				var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
			})();
		</script>
		<link href="../../style/css/default.css" rel="stylesheet" type="text/css">
		<title></title>
	</head>
	<body>
		<div class="header" id="header1">
			<!--<img src = "style/images/qctool.png" class = "header_img" ></img>-->
			<div class="header_text">
				qctool v2
			</div>
			<div class="header_subtext">
				A tool for quality control and analysis of gwas datasets.
			</div>
		</div>

		<section class="main_text">
			<h2>Using qctool in a pipeline</h2>
				<div class="task">
					<div class="task_name">
						Outputting data to stdout
					</div>
					<div class="task_notes">
						Using "-" as an output file specifier will output data to standard output, that can then be redirected
						or piped into other commands.  For example:
					</div>
					<div class="task_command_line">
						$ qctool -g example.bgen -og -
					</div>
					<div class="task_notes">
						This outputs genotype data to stdout.
						(By default output is in VCF format; use the <code>-ofiletype</code> option to alter this).
					</div>
					<div class="task_command_line">
						$ qctool -g example.bgen -snp-stats -osnp -
					</div>
					<div class="task_notes">
						This outputs SNP summary statistics to standard output.
					</div>
				</div>
				<div class="task">
					<div class="task_name">
						Inputting data from stdin
					</div>
					<div class="task_notes">
						QCTOOL has limited support for reading data from stdin.  To do this, specify "-" as the argument to <code>-g</code>.  E.g.:
					</div>
					<div class="task_command_line">
						$ cat file.bgen | qctool -g - -filetype bgen [other options]
					</div>
					<div class="task_notes">
					  <p>
							This functionality is currently limited to input in BGEN format.
							You must specify <code>-filetype bgen</code> to tell QCTOOL a BGEN file is being input.
						</p>
					</div>
				</div>
				<div class="task">
					<div class="task_name">
						Pipelining QCTOOL with other tools
					</div>
					<div class="task_notes">
						As an example, the following command uses <a href="http://bitbucket.org/gavinband/bgen">bgenix</a> with QCTOOL to
						compute snp summary statistics from a subset of a BGEN file, and view the result on the fly using <a href="https://en.wikipedia.org/wiki/Less_(Unix)"><code>less</code><a>.
					</div>
					<div class="task_command_line">
						$ bgenix -g file.bgen -range 11:3500000-6500000 | qctool -g - -filetype bgen -snp-stats -osnp - | less -S
					</div>
				</div>
		</section>

		<nav class="button_bar">
			<div class="nav_button" name="overview">
				<a href="../../index.html">overview</a>
			</div>
			<div class="nav_button" name="documentation">
				documentation
				<ul class="menu">
					<li><a href="../../documentation/diagram_of_options.html">Diagram of options</a></li>
					<li><a href="../../documentation/alphabetical_options.html">List of options</a></li>
					<li><a href="../../documentation/summary_statistics.html">Computing summary statistics</a></li>
					<li><a href="../../documentation/conversions.html">Converting between formats</a></li>
					<li><a href="../../documentation/mapping_and_aligning.html">Remapping and aligning data</a></li>
					<li><a href="../../documentation/annotations.html">Annotating data</a></li>
				</ul>
			</div>
			<div class="nav_button" name="tutorial">
				examples
				<ul class="menu">
					<li><a href="../../documentation/examples/basic.html">Basic usage</a></li>
					<li><a href="../../documentation/examples/converting.html">Converting file formats</a></li>
					<li><a href="../../documentation/examples/sorting.html">Sorting and reordering data</a></li>
					<li><a href="../../documentation/examples/combining.html">Combining datasets</a></li>
					<li><a href="../../documentation/examples/merging.html">Merging in variants</a></li>
					<li><a href="../../documentation/examples/altering_id_data.html">Altering variant identifying data</a></li>
					<li><a href="../../documentation/examples/pipelining.html">Using qctool in a pipeline</a></li>
					<li><a href="../../documentation/examples/filtering_samples.html">Filtering samples</a></li>
<li><a href="../../documentation/examples/filtering_variants.html">Filtering variants</a></li>
					<li><a href="../../documentation/examples/summary_statistics.html">Computing summary statistics</a></li>
					<li><a href="../../documentation/examples/adding_to_sample_file.html">Adding data to a sample file</a></li>
<li><a href="../../documentation/examples/computing_pcs.html">Computing principal components</a></li>
				</ul>
			</div>
			<div class="nav_button" name="file_formats">
				file formats
				<ul class="menu">
					<a href="../../documentation/genotype_file_formats.html">Genotype file formats</a>
					<a href="../../documentation/sample_file_formats.html">Sample file formats</a>
					<a href="../../documentation/summary_file_formats.html">Summary statistic file formats</a>
			</div>
			<div class="nav_button" name="download">
				<a href="../../documentation/download.html">download</a>
			</div>
		</nav>
		
	</body>
</html>
